File: matrix-traversals/dfs.py
def dfs(grid: List[List[int]], i: int, j: int, visited: Set[List[int]]):
    if i < 0 or j < 0 or i >= len(grid) or \
    j >= len(grid[0]) or (i, j) in visited: return
    
    visited.add((i, j))
    dfs(grid, i + 1, j, visited)
    dfs(grid, i - 1, j, visited)
    dfs(grid, i, j + 1, visited)
    dfs(grid, i, j - 1, visited)

File: matrix-traversals/dfs.java
void dfs(int[][] grid, int i, int j, boolean[][] visited) {
    if (i < 0 || j < 0 || i >= grid.length || 
        j >= grid[0].length || visited[i][j]) return;
    visited[i][j] = true;
    dfs(grid, i + 1, j, visited);
    dfs(grid, i - 1, j, visited);
    dfs(grid, i, j + 1, visited);
    dfs(grid, i, j - 1, visited);
}